lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +14,241 @@
 //! with context sensitive (e.g. indentation-aware) syntax.
 
-pub use enso_prelude as prelude;
+// TODO [AA] Logging https://github.com/enso-org/ide/blob/main/src/rust/ide/src/model/execution_context/synchronized.rs#L45
 
+use crate::prelude::*;
 
+use lazy_reader::LazyReader;
+use crate::group::GroupRegistry;
 
 pub mod automata;
-pub mod group;
 pub mod data;
+pub mod group;
+
+#[allow(missing_docs)]
+pub mod prelude {
+    pub use enso_prelude::*;
+}
 
 
 
-// ===============
-// === Flexer ====
-// ===============
+// =================
+// === Constants ===
+// =================
+
+mod constants {
+    /// The number of 'frames' to reserve in the state stack.
+    pub const STATE_STACK_RESERVATION:usize = 1024;
+    /// The size of the output buffer (in tokens) to reserve.
+    pub const OUTPUT_BUFFER_RESERVATION:usize = 1024;
+}
+
+
+
+// ==============
+// === Flexer ===
+// ==============
 
 /// The flexer is an engine for generating lexers.
 ///
 /// Akin to flex and other lexer generators, it is given a definition as a series of rules from
 /// which it then generates code for a highly optimised lexer implemented on top of a
 /// [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
-pub trait Flexer {
+///
+/// Lexers defined using the flexer work on a stack of _states_, where a state is represented by a
+/// [`crate::group::Group`]. Being in a given state (represented below by the top of the
+/// `state_stack`) means that the flexer can match a certain set of rules associated with that
+/// state. The user may cause the lexer to transition between states by pushing and popping states
+/// on the stack, thus allowing a much more flexible lexing engine than pure regular grammars.
+#[derive(Clone,Debug)]
+pub struct Flexer<Definition,Output,Reader> {
+    /// The stack of states that are active during lexer execution.
+    pub state_stack: Vec<group::Identifier>,
+    /// A reader for the input.
+    pub reader: Reader,
+    /// The result of the current stage of the DFA.
+    pub status: StageStatus,
+    /// The tokens that have been lexed.
+    pub output: Vec<Output>,
+    /// The text of the current match of the lexer.
+    pub current_match: String,
+    /// The definition of the user-provided state for the lexer.
+    definition: Definition,
+}
+
+impl<Definition,Output,Reader> Flexer<Definition,Output,Reader>
+where Definition:State, Reader:LazyReader {
+    /// Creates a new lexer instance.
+    pub fn new(mut reader:Reader) -> Flexer<Definition,Output,Reader> {
+        let mut state_stack = Vec::default();
+        state_stack.reserve(constants::STATE_STACK_RESERVATION);
+        let status     = default();
+        let mut output = Vec::default();
+        output.reserve(constants::OUTPUT_BUFFER_RESERVATION);
+        let definition       = Definition::new(&mut reader);
+        let initial_state_id = definition.initial_state();
+        state_stack.push(initial_state_id);
```

Comment:
why we dont put all lets first and all actions later? IT would make the code more clear and more easy to read imo. First we can define all dependencies and then all actions (just like "where" block in Haskell)